MovingBlocks/DestinationSol
# [Work In Progress] Gestalt integration

**Important:** This PR contains significant amounts of whitespace-only changes caused by IntelliJ autoformatting. Append `?w=1` to the end of the diff view URL to display non-whitespace changes only.
### Contains

A work-in progress, proof-of-concept integration of Destination Sol with [gestalt-asset-core](https://github.com/MovingBlocks/gestalt/tree/develop/gestalt-asset-core), as initially suggested by [this forum post](http://forum.terasology.org/threads/hello-world.1446/#post-12892).
##### Features and TODO list
- [X] A ClasspathModule providing access to core functionality and resources, with required metadata being stored in a module.info file and parsed using a specialized `ModuleManager` class.
- [X] A generalized `AssetHelper` class registering asset types within a `ModuleEnvironment` containing the core ClasspathModule (and, potentially, external ArchiveModules), as well as providing generalized wrappers to simple asset operations (get asset of type, get all assets of type).
- [X] `OggSound` asset type (including implementations for `AssetData`, `Asset`, `AbstractAssetFileFormat`, `AbstractAssetAlterationFileFormat`). Acts as a data wrapper for:
  - a `gdx.audio.Sound` object - obtained from a `FileHandle` generated by reconstructing the path of the input asset file in `OggSoundFileFormat`;
  - configuration parameters fetched from a metadata file (`loopTime` / `baseVolume`);
  - a parameter programmatically defined configuration parameter (`basePitch`).
- [x] Asset-based sound manager class - a refactored version of [`SoundManager`](https://github.com/MovingBlocks/DestinationSol/blob/develop/main/src/org/destinationsol/game/sound/SoundManager.java) that utilises the OggSound class to implement cleaner sound fetching and improved playback code.
- [x] A refactor for the game to support said manager class
- [ ] Texture asset type and file format. Should read the game's textures and metadata atlas files (with sol.atlas being split into different files corresponding to each texture file).
- [ ] Asset-based texture manager class - a refactored version of [`AtlasTextureProvider`](https://github.com/MovingBlocks/DestinationSol/blob/develop/main/src/org/destinationsol/AtlasTextureProvider.java).
- [ ] Other asset types - complete list up for discussion.
- [ ] Test the integration's compatibility with Android and GWT facades.
### How to test

Launch the game from source! An asset fetching test in SolApplication should output a list of available `OggSound` and `OggMusic` Urns - verify this.

The sound system is now fully integrated into the game, with all the sounds being fetched from an `OggSoundManager`, so simply start a new game and play around with things - the sounds currently being played will be logged to the console.
